### PR TITLE
Consume snippet generator as a tool

### LINKED ISSUE #10997: [FEATURE REQ] Automatically run Export-API.ps1 in Pull Requests
Would it be possible to automagically run Export-Api.ps1 script in Pull Requests and in CI pipelines?
It's quite annoying to do this manually. If one forgets to do that then it's waste of time and unnecessary PR iteration.

### PR DESCRIPTION
**Pull request by chidozieononiwu**
- Consume snippet generator as a tool publisehd to the azure-sdk-tools feed.
- Depends on https://github.com/Azure/azure-sdk-tools/pull/1834
- Should resolve https://github.com/Azure/azure-sdk-for-net/issues/10997 and https://github.com/Azure/azure-sdk-for-net/issues/13679

### CODE REVIEW

File: NuGet.Config [hunk 1]
```diff
@@ -10,6 +10,7 @@
       Used for azure-sdk-tools repo until issue https://github.com/Azure/azure-sdk-tools/issues/1329 is addressed
     -->
     <add key="azure-sdk-tools" value="https://azuresdkartifacts.blob.core.windows.net/azure-sdk-tools/index.json" />
+    <add key="azure-sdk-tools@DevOps" value="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-tools/nuget/v3/index.json" />
```
Comment (weshaggard) [inline NuGet.Config:13]: Is this needed yet? Can we eliminate the other feed? I'm trying to keep us down to a minimal set of feeds.

File: eng/scripts/Update-Snippets.ps1 [hunk 2]
```diff
@@ -19,8 +18,20 @@ if ($ServiceDirectory -and ($ServiceDirectory -ne "*")) {
 
 if (-not (Test-Path env:TF_BUILD)) { $StrictMode = $true }
 
-if($StrictMode) {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -sm -c Release }
-} else {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -c Release }
+nuget install SnippetGenerator -Source azure-sdk-tools@DevOps -PreRelease -OutputDirectory .
+$generatorAssembly = Resolve-Path -Path ".\SnippetGenerator*\tools\**\SnippetGenerator.dll"
```
Comment (weshaggard) [inline eng/scripts/Update-Snippets.ps1:22]: Why do we need to resolve the path? After you install a global tool you should be able to use it from the path.

File: eng/scripts/Update-Snippets.ps1 [hunk 3]
```diff
@@ -19,8 +18,20 @@ if ($ServiceDirectory -and ($ServiceDirectory -ne "*")) {
 
 if (-not (Test-Path env:TF_BUILD)) { $StrictMode = $true }
 
-if($StrictMode) {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -sm -c Release }
-} else {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -c Release }
+nuget install SnippetGenerator -Source azure-sdk-tools@DevOps -PreRelease -OutputDirectory .
```
Comment (weshaggard) [inline eng/scripts/Update-Snippets.ps1:21]: Do we have plans to use this script outside of the tools repo? If so we need to not have a dependency on a feed name but instead we probably need to give the feed path here.

File: eng/scripts/Update-Snippets.ps1 (hunk 2 above)
Comment (chidozieononiwu) [reply eng/scripts/Update-Snippets.ps1:22]: Am actually installingit directly to the current directory then cleaning up after.

File: NuGet.Config (hunk 1 above)
Comment (chidozieononiwu) [reply NuGet.Config:13]: I removed the new feed and added it directly as the source in the Update-Snippet Script.

File: eng/scripts/Update-Snippets.ps1 [hunk 4]
```diff
@@ -19,8 +18,22 @@ if ($ServiceDirectory -and ($ServiceDirectory -ne "*")) {
 
 if (-not (Test-Path env:TF_BUILD)) { $StrictMode = $true }
 
-if($StrictMode) {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -sm -c Release }
-} else {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -c Release }
+nuget install SnippetGenerator `
```
Comment (weshaggard) [inline eng/scripts/Update-Snippets.ps1:21]: This is going to fail on non-windows. Don't you want to use `dotnet tool install`?

File: eng/scripts/Update-Snippets.ps1 (hunk 2 above)
Comment (weshaggard) [reply eng/scripts/Update-Snippets.ps1:24]: I understand you are installing it locally but you should still be able to run it without needing to glob and find the path like you are doing here.

Comment (weshaggard) [reply eng/scripts/Update-Snippets.ps1:24]: Is there any reason the tool doesn't have a exe in the install location? Or any reason you cannot do `dotnet run .\SnippetGenerator`

Comment (weshaggard) [reply eng/scripts/Update-Snippets.ps1:24]: Looks like you didn't package this as a tools package see https://github.com/Azure/azure-sdk-tools/blob/main/tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Azure.Sdk.Tools.PipelineGenerator.csproj#L7 which is what you need to get a standard dotnet tools package.

File: eng/scripts/Update-Snippets.ps1 [hunk 5]
```diff
@@ -19,8 +18,20 @@ if ($ServiceDirectory -and ($ServiceDirectory -ne "*")) {
 
 if (-not (Test-Path env:TF_BUILD)) { $StrictMode = $true }
 
-if($StrictMode) {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -sm -c Release }
-} else {
-    Resolve-Path "$root" | %{ dotnet run -p $generatorProject -b "$_" -c Release }
+dotnet tool install Azure.Sdk.Tools.SnippetGenerator `
+-g --add-source "https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-tools/nuget/v3/index.json" `
```
Comment (weshaggard) [inline eng/scripts/Update-Snippets.ps1:22]: I assume the plan is to use the azure-sdk-for-net feed once you make your other changes.

**Review by weshaggard — APPROVED** [review]
As we chatted I like that this is a tool now but there is still some risk of a race condition from installing the tool globally but I think that risk is minimal compared to compiling the project. 
We might want to potentially consider pinning the version instead of using a * version on the install.

Another thing to consider is I suspect that we are running these codechecks multiple times for each service directory, once per project, so we are likely doing more work then we need to but as long as the work doesn't take too long it probably doesn't matter.

**Pull request merged**